pkg/hpsf/hpsfTypes.go

```diff
@@ -526,6 +559,117 @@ func (h *HPSF) GetStartComponents() []*Component {
 	return startComps
 }
 
+// for a given signal type, generate a list of components that are sources of connections but not destinations
+// of connections. This is used to find the start components of a pipeline.
+func (h *HPSF) GetSourceComponentsFor(signalType ConnectionType) []*Component {
+	sourceComps := make([]*Component, 0)
+	// make a map of all components that are destinations of connections for the given signal type
+	destinations := make(map[string]bool)
+	for _, conn := range h.Connections {
+		if conn.Source.Type == signalType {
+			destinations[conn.Destination.Component] = true
+		}
+	}
+	for _, c := range h.Components {
+		// if the component is a source of a connection and not a destination of a connection, add it to the list
+		if h.isSourceComponent(c, signalType) && !destinations[c.Name] {
+			sourceComps = append(sourceComps, c)
+		}
+	}
+	// sort the components by name
+	slices.SortFunc(sourceComps, func(a, b *Component) int {
+		return strings.Compare(a.Name, b.Name)
+	})
+	return sourceComps
+}
+
+func (h *HPSF) getComponent(name string) *Component {
+	// find the component with the given name
+	for _, c := range h.Components {
+		if c.Name == name {
+			return c
+		}
+	}
+	return nil
+}
+
+func (h *HPSF) isSourceComponent(c *Component, signalType ConnectionType) bool {
+	// check if the component is a source of a connection of this signal type
+	for _, conn := range h.Connections {
+		if conn.Source.Component == c.Name && conn.Source.Type == signalType {
+			return true
+		}
+	}
+	return false
+}
+
+type PipelineWithSignalType struct {
+	SignalType ConnectionType
+	Pipeline   []*Component
+}
+
+func (p PipelineWithSignalType) GetID() string {
+	// return the ID of the pipeline, which is a hash of the names of components in its pipeline
+	// and the signal type, truncated to 4 characters.
+	buf := bytes.Buffer{}
+	for _, comp := range p.Pipeline {
+		buf.WriteString(comp.GetSafeName())
+	}
+	buf.WriteString(string(p.SignalType))
+	hash := metro.Hash64(buf.Bytes(), 0x234da488) // use a fixed seed for reproducibility
+	shash := strconv.FormatUint(hash, 16)
+	return shash[len(shash)-4:] // return the last 4 characters of the hash
```

Comment:
What is the probability of two pipelines generating a hash where the last 4 characters match and would clash?

Reply:
There are 65536 possible 4-letter hashes. There is a one in 10,000 chance of collision when there are 5 pipelines in the mix. Even if there are 500 pipelines it's still well under 1%. 

HPSF works pretty hard to make its generated configs both legible and idiomatic by humans. Long hex strings are not something normal humans are comfortable with. 

I changed it to 6-character, with a dash, which drops the collision probability to 1 in a million for 4 pipelines, 1 in 10,000 for 500 pipelines.